src/object/srv_obj.c

```diff
@@ -954,6 +950,24 @@ obj_local_rw(crt_rpc_t *rpc, struct ds_cont_hdl *cont_hdl,
 		goto out;
 	}
 
+	if (obj_rpc_is_fetch(rpc) && !size_fetch) {
+		obj_fetch_csum_init(cont_hdl, orw, orwo);
+
+		obj_fetch_csums_link(orw, orwo);
+
+		rc = vos_fetch_csum(ioh,
```

Comment:
So we fetch data from NVMe to DMA buffer first in above bio_iod_prep(), then yield (bio_iod_prep() will yield on DMA transfer), then we fetch csum, the data could have been changed by new updates while yielding.

Not sure if I fully understand the csum implementation, but I think we need to ensure the atomicity of 'fetching csum & acquiring data address on SCM or NVMe'.

Reply:
Not exactly and maybe we need a better name than "vos_fetch_csum". The checksums are actually "fetched" during vos_fetch_begin but stored internally in the vos_io_context. vos_fetch_csum either copies the checksums to the io descriptor or calculates new ones if extents arn't chunk aligned.

Reply:
we do need a better name, and comment for the function as well. 

> So we fetch data from NVMe to DMA buffer first in above bio_iod_prep(), then yield (bio_iod_prep() will yield on DMA transfer), then we fetch csum, the data could have been changed by new updates while yielding.
> 
> Not sure if I fully understand the csum implementation, but I think we need to ensure the atomicity of 'fetching csum & acquiring data address on SCM or NVMe'.

I don't think there is yield at here, but we could do this in the future because we might offload the checksum compute to a different xstream and yield, the  aggregation merge extent at the background?